editor/src/clj/editor/console.clj

```diff
@@ -65,33 +90,189 @@
 (defn clear-console!
   "Clear the console. Callable from a background thread."
   []
-  (reset! pending-atom {:clear? true :entries PersistentQueue/EMPTY}))
+  (swap! pending-atom assoc :clear true :entries [] :index 0))
+
+(defn- consume-entries-in-state [state ^long n]
+  (let [{:keys [^long index entries]} state]
+    (assoc state :index (min (count entries) (+ index n)) :clear false)))
 
-(defn- pop-n [coll ^long n]
-  (let [max-n (min n (count coll))]
-    (loop [i 0
-           xs coll]
-      (if (= max-n i)
-        xs
-        (recur (inc i) (pop xs))))))
+(defn- get-entries-from-state [state ^long n]
+  (let [{:keys [^long index entries entry-pred]} state
+        unfiltered (subvec entries index (min (count entries) (+ index n)))]
+    (if entry-pred
+      (filterv entry-pred unfiltered)
+      unfiltered)))
 
 (defn- dequeue-pending! [n]
-  (let [batch (first (swap-vals!
-                       pending-atom
-                       (fn [{:keys [entries]}]
-                         {:clear? false
-                          :entries (pop-n entries n)})))]
-    (update batch :entries #(take n %))))
+  (let [old-state (first (swap-vals! pending-atom consume-entries-in-state n))]
+    (assoc old-state :entries (get-entries-from-state old-state n))))
+
+(defn- compile-entry-predicate [filter-set]
+  (letfn [(make-inclusion-pred [inclusion]
+            (fn inclusion-pred [entry]
+              (string/includes? (entry 1) inclusion)))
+          (make-exclusion-pred [exclusion-with-prefix]
+            (let [exclusion (subs exclusion-with-prefix 1)]
+              (fn exclusion-pred [entry]
+                (not (string/includes? (entry 1) exclusion)))))
+          (combine-preds [pred-combinator preds]
+            (case (count preds)
+              0 nil
+              1 (first preds)
+              (apply pred-combinator preds)))]
+    (let [[exclusions inclusions] (coll/separate-by #(and (<= 2 (count %))
+                                                          (string/starts-with? % "!"))
```

Comment:
What role does `(<= 2 (count %)` play here? Isn't the `starts-with?` check sufficient?

Reply:
I wanted to ensure that there are at least 2 characters so it's not a single exclamation mark. After applying `(subs "!" 1)` we will get an empty string which will make us exclude every line since every line contains an empty string %)

Reply:
Ah, makes sense!

Reply:
Yes, which is weird, but less weird than excluding everything with 😄